modules/optimization/src/optimizationreporters/ParameterMeshOptimization.C

```diff
@@ -0,0 +1,182 @@
+#include "ParameterMeshOptimization.h"
+
+#include "AddVariableAction.h"
+#include "ParameterMesh.h"
+#include "libmesh/string_to_enum.h"
+
+registerMooseObject("OptimizationApp", ParameterMeshOptimization);
+
+InputParameters
+ParameterMeshOptimization::validParams()
+{
+  InputParameters params = OptimizationReporterBase::validParams();
+  params.addClassDescription(
+      "Computes objective function, gradient and contains reporters for communicating between "
+      "optimizeSolve and subapps using mesh-based parameter definition.");
+
+  params.addRequiredParam<std::vector<ReporterValueName>>(
+      "parameter_names", "List of parameter names, one for each group of parameters.");
+  params.addRequiredParam<std::vector<FileName>>(
+      "parameter_meshes", "Exodus file containing meshes describing parameters.");
+
+  const auto family = AddVariableAction::getNonlinearVariableFamilies();
+  MultiMooseEnum families(family.getRawNames(), "LAGRANGE");
+  params.addParam<MultiMooseEnum>(
+      "parameter_families",
+      families,
+      "Specifies the family of FE shape functions for each parameter. If a single value is "
+      "specified, then that value is used for all parameters.");
+  const auto order = AddVariableAction::getNonlinearVariableOrders();
+  MultiMooseEnum orders(order.getRawNames(), "FIRST");
+  params.addParam<MultiMooseEnum>(
+      "parameter_orders",
+      orders,
+      "Specifies the order of FE shape functions for each parameter. If a single value is "
+      "specified, then that value is used for all parameters.");
+
+  params.addParam<unsigned int>(
+      "num_parameter_times", 1, "The number of time points the parameters represent.");
+
+  params.addParam<std::vector<Real>>("initial_condition",
+                                     std::vector<Real>(),
+                                     "Spatially constant initial condition for each parameter.");
+  params.addParam<std::vector<Real>>(
+      "lower_bounds", std::vector<Real>(), "Spatially constant lower bound for each parameter.");
+  params.addParam<std::vector<Real>>(
+      "upper_bounds", std::vector<Real>(), "Spatially constant upper bound for each parameter.");
+
+  return params;
+}
+
+ParameterMeshOptimization::ParameterMeshOptimization(const InputParameters & parameters)
+  : OptimizationReporterBase(parameters),
+    _parameter_names(getParam<std::vector<ReporterValueName>>("parameter_names")),
+    _nparams(_parameter_names.size()),
+    _nvalues(_nparams),
+    _parameters(_nparams),
+    _gradients(_nparams),
+    _lower_bounds(getParam<std::vector<Real>>("lower_bounds")),
+    _upper_bounds(getParam<std::vector<Real>>("upper_bounds"))
+{
+  const auto & meshes = getParam<std::vector<FileName>>("parameter_meshes");
+  const auto & families = getParam<MultiMooseEnum>("parameter_families");
+  const auto & orders = getParam<MultiMooseEnum>("parameter_orders");
+  const auto & initial_condition = getParam<std::vector<Real>>("initial_condition");
```

Comment:
Would you mind if we push this down the road a bit? Parameter value come in the form of vector reporters, it would be quite wasteful to copy it into a solution at this point. These reporter values are restartable data, so it should work fine with restart. I need to think more about initial conditions, I don't think that part is required for the purposes of this PR.

Reply:
agree